### PR TITLE
[Feature] Add ENet

### PR DESCRIPTION
**Pull request by ETTR123**

### CODE REVIEW

File: paddleseg/models/enet.py [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+    Main branch:
+    1. Shortcut connection.
+    Extension branch:
+    1. 1x1 convolution which decreases the number of channels by
+        ``internal_ratio``, also called a projection;
+    2. regular, dilated or asymmetric convolution;
+    3. 1x1 convolution which increases the number of channels back to
+        ``channels``, also called an expansion;
+    4. dropout as a regularizer.
+    Args:
```
Comment (shiyutang) [inline paddleseg/models/enet.py:306]: Plz add appropriate blank lines to the note, here and elsewhere.

File: paddleseg/models/enet.py [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+                      stride=1,
+                      bias_attr=bias), layers.SyncBatchNorm(channels),
+            activation())
+
+        self.ext_regul = nn.Dropout2D(p=dropout_prob)
+
+        self.out_activation = activation()
+
+    def forward(self, x):
+        # Main branch shortcut
```
Comment (shiyutang) [inline paddleseg/models/enet.py]: Remove this annotation plz, the code is clear by itself

File: paddleseg/models/enet.py [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+                      stride=1,
+                      bias_attr=bias), layers.SyncBatchNorm(out_channels),
+            activation())
+
+        self.ext_regul = nn.Dropout2D(p=dropout_prob)
+
+        self.out_activation = activation()
+
+    def forward(self, x):
+        # Main branch shortcut
```
Comment (shiyutang) [inline paddleseg/models/enet.py]: remove this plz, here and elsewhere.

File: configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      saturation_range: 0.4
+    - type: Normalize
+  mode: train
+
+val_dataset:
+  type: Cityscapes
+  dataset_root: data/cityscapes
+  transforms:
+    - type: Normalize
+  mode: val
```
Comment (shiyutang) [inline configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml]: You can inherit the cityscapes.yml which have the same valset and loss config as yours. Just set the __base__ as following and you can refer to this: https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.3/configs/deeplabv3/deeplabv3_resnet50_os8_cityscapes_1024x512_80k.yml
```_base_: '../_base_/cityscapes.yml'```

Comment (shiyutang) [reply configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml]: When you set this, it would be great if you adjust the order of configs like that config as well.

**Review by shiyutang — CHANGES_REQUESTED** [review]

File: paddleseg/models/enet.py (hunk 1 above)
Comment (ETTR123) [reply paddleseg/models/enet.py:306]: Done.

File: paddleseg/models/enet.py (hunk 2 above)
Comment (ETTR123) [reply paddleseg/models/enet.py]: Done.

File: configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml (hunk 4 above)
Comment (ETTR123) [reply configs/enet/enet_cityscapes_1024x512_adam_0.002_80k.yml]: Done.

**Review by shiyutang — COMMENTED** [review]
Good job, I will validate it and merge if the result is solid.

File: paddleseg/core/train.py [hunk 5]
```diff
@@ -314,7 +314,7 @@ def train(model,
             batch_start = time.time()
 
     # Calculate flops.
-    if local_rank == 0:
+    if local_rank == 0 and model.__class__.__name__ != 'ENet':
```
Comment (shiyutang) [inline paddleseg/core/train.py]: this line can be removed on nightly build paddle-dev.

**Review by shiyutang — COMMENTED** [review]
The code looks good to me. It needs to be further validated on the paddle-dev and on the validation set.

**Comment by shiyutang** [comment]
Validate on the dataset
image

**Review by shiyutang — APPROVED** [review]
